ops/script/rename_whl.py

```diff
@@ -0,0 +1,62 @@
+import argparse
+import pathlib
+
+
+def main(args: argparse.Namespace) -> None:
+    wheel_path = pathlib.Path(args.wheel_path).expanduser().resolve()
+    if not wheel_path.exists():
+        raise ValueError(f"Wheel cannot be found at path {wheel_path}")
+    if not wheel_path.is_file():
+        raise ValueError(f"Path {wheel_path} is not a valid file")
+    wheel_dir, wheel_name = wheel_path.parent, wheel_path.name
+
+    tokens = wheel_name.split("-")
+    assert len(tokens) == 5
+    version = tokens[1].split("+")[0]
+    keywords = {
+        "pkg_name": tokens[0],
+        "version": version,
+        "commit_id": args.commit_hash,
+        "platform_tag": args.platform_tag,
+    }
+    new_wheel_name = (
+        "{pkg_name}-{version}+{commit_id}-py3-none-{platform_tag}.whl".format(
+            **keywords
+        )
+    )
+    new_wheel_path = wheel_dir / new_wheel_name
+    print(f"Renaming {wheel_name} to {new_wheel_name}...")
+    if new_wheel_name == wheel_name:
+        print("Skipping, as the old name is identical to the new name.")
+    else:
+        if new_wheel_path.is_file():
+            new_wheel_path.unlink()
+        wheel_path.rename(new_wheel_path)
+
+    filesize = new_wheel_path.stat().st_size / 1024 / 1024  # MiB
+    print(f"Wheel size: {filesize:.2f} MiB")
+
+    if filesize > 300:
+        raise RuntimeError(
+            f"Limit of wheel size set by PyPI is exceeded. {new_wheel_name}: {filesize:.2f} MiB"
+        )
```

Comment:
Thanks for your recommendation. I agree with you that this script would better be removed.

Reply:
On a second thought, I'd like to retain `rename_whl.py`, to add the local version label (`+commit_hash`) to the version string. The reason is that we re-use nightly artifacts to make release of stable releases. Changing the version string to add the local version label would change the metadata inside, and we don't want to use `3.0.1+commit_hash` as the version string for version 3.0.1. Basically, I want to be able to add the local version label (`+commit_hash`) to the file name only without changing the metadata inside the wheel.

Reply:
On a second thought, I'd like to retain `rename_whl.py`. Currently, we add the commit hash to the file name name without having to also update the metadata. So we can download `xgboost-2.1.3+600be4dbb54ec5cda4cb61a2bb8fc4bce6f3fc58-py3-none-win_amd64.whl` (a nightly build from branch `release_2.1.0`) and it will produce the version string `2.1.3`. If we rename this wheel to `xgboost-2.1.3-py3-none-win_amd64.whl`, we can publish it as an artifact for the stable version 2.1.3.

Whereas in the new method, where we overwrite the version string in `pyproject.toml`, we have to re-build a new wheel in order to make a stable release; we cannot re-use a nightly artifact because its metadata would still contain the version string even after renaming the file name.

Reply:
In general, I don't recommend having the filename for the wheel imply a different version than the package metadata... those are generally assumed to be consistent by tools, I think (though I can't think of a specific example where this would cause a problem).

If it's important to preserve the give commit information, what about adding `/{commit_sha}/` to the key on S3 instead? Or is there some specific reason it needs to be recoverable from the filename specifically?

Reply:
I decided to add `commit_sha` to the S3 key instead, so that we don't have to rename the wheel.